File: common/rhino_common/text_manipulate.js
// ------------------------------------------------------------------
//
//	text_manipulate.js
//
//						Oct/24/2011
//
// ------------------------------------------------------------------
function text_read_proc (file_in)
{
	var dict_aa = new Object ();

	var lines_in = file_read_proc (file_in);

	for (var it=0; it< lines_in.length; it++)
		{
		var rr = String.split (lines_in[it],"\t");

		dict_aa[rr[0]] = {"name": rr[1],"population": rr[2], "date_mod": rr[3]};
		}
	
	return	dict_aa;
}

// ------------------------------------------------------------------
function dict_display_proc (dict_aa)
{
	for (var id in dict_aa)
		{
		if ((id != '_id') && (id != '_rev'))
			{
		var out_str = id + "\t";
		out_str += dict_aa[id]["name"] + "\t";
		out_str += dict_aa[id]["population"] + "\t";
		out_str += dict_aa[id]["date_mod"];
		print (out_str);
			}
		}
	
}

// ---------------------------------------------------------------
function text_write_proc (file_out,dict_aa)
{
	var out_str = "";

	for (var id in dict_aa)
		{
		out_str += id + "\t";
		out_str += dict_aa[id]["name"] + "\t";
		out_str += dict_aa[id]["population"] + "\t";
		out_str += dict_aa[id]["date_mod"] + "\n";
		}

	string_write_proc (file_out,out_str);
}


// ------------------------------------------------------------------
function dict_update_proc (dict_in,id,population)
{
	dict_in[id]["population"] = population;
	dict_in[id]["date_mod"] = get_current_date_proc ();

	return dict_in;
}

// ------------------------------------------------------------------
function get_current_date_proc ()
{
	var today = new Date ();
	var ddx = (1900 + today.getYear ()) + "-" + (today.getMonth () +1)
	ddx += "-" + today.getDate ()

	return ddx;
}

// ------------------------------------------------------------------
function dict_append_proc (dict_aa,id_in,name_in,population_in,date_mod_in)
{
	unit_aa = {};
	unit_aa['name'] = name_in;
	unit_aa['population'] = population_in;
	unit_aa['date_mod'] = date_mod_in;

	dict_aa[id_in] = unit_aa;

	return	dict_aa;
}

// ------------------------------------------------------------------
function gen_record (id_in,name_in,population_in,date_mod_in)
{
	unit_aa = {};
	unit_aa['id'] = id_in;
	unit_aa['name'] = name_in;
	unit_aa['population'] = population_in;
	unit_aa['date_mod'] = date_mod_in;

	return	unit_aa;
}

// ------------------------------------------------------------------
function csv_read_proc (file_in)
{
	var dict_aa = new Object ();

	var lines_in = file_read_proc (file_in);

	for (var it=0; it< lines_in.length; it++)
		{
		var rr = String.split (lines_in[it],",");

		dict_aa[rr[0]] = {"name": rr[1],"population": rr[2], "date_mod": rr[3]};
		}
	
	return	dict_aa;
}

// ------------------------------------------------------------------
function csv_write_proc (file_out,dict_aa)
{
	var out_str = "";

	var delimit = ",";

	for (var id in dict_aa)
		{
		out_str += id + delimit;
		out_str += dict_aa[id]["name"] + delimit;
		out_str += dict_aa[id]["population"] + delimit;
		out_str += dict_aa[id]["date_mod"] + "\n";
		}

	string_write_proc (file_out,out_str);
}


// ------------------------------------------------------------------

File: common/lua_common/text_manipulate.lua
-- ----------------------------------------------------
--	text_manipulate.lua
--
--					Dec/15/2014
-- ----------------------------------------------------
function split_proc (str, del)
	p, nrep = str:gsub("%s*"..del.."%s*", "")
	return { str:match((("%s*(.-)%s*"..del.."%s*"):rep(nrep).."(.*)")) } 
end 
-- ----------------------------------------------------
function join_proc (doc_aa,id,delimit)
	name = doc_aa[id]["name"]
	population = doc_aa[id]["population"]
	date_mod = doc_aa[id]["date_mod"]
	out_str = id .. delimit
		.. name .. delimit
		.. population .. delimit
		.. date_mod
	return	out_str
end
-- ----------------------------------------------------
function get_keys_sorted_proc (tbl)
	local keys = {}
	for key in pairs(tbl) do
		table.insert(keys, key)
	end

	table.sort (keys)

	return keys
end

-- ----------------------------------------------------
function dict_display_proc (dict_in)
	keys = get_keys_sorted_proc (dict_in)
	for num,value in pairs (keys)
		do
			if ((value ~= "_id") and (value ~= "_rev")) then
			if (dict_in[value]) then
				out_str = join_proc (dict_in,value,"\t")
				print (out_str)
			end
			end
		end
end
-- ----------------------------------------------------
function text_read_proc (file_in)
	fp_in, msg=io.open (file_in,"r")

	doc_aa = {}

	if (fp_in) then
		tt = fp_in:read ("*a")
		fp_in:close ()
		lines=split_proc (tt, "\n")
		for it=1 , #lines do
			rr=split_proc (lines[it], "\t")
			if (2 < #rr) then
				doc_aa[rr[1]] =
				 {name= rr[2],population=rr[3],date_mod=rr[4]}
			end
		end
	else
		print (msg)
	end

	return	doc_aa
end

-- ----------------------------------------------------
function text_write_proc (file_out,dict_in)
	io.output (file_out)
	for str_key in pairs (dict_in)
		do
			if (dict_in[str_key]) then
				out_str = join_proc (dict_in,str_key,"\t") .. "\n"
				io.write (out_str)
			end
		end
end
-- ----------------------------------------------------
function dict_update_proc (dict_in,id_in,population_in)
	str_key = tostring (id_in)

	if (dict_in[str_key]) then
		dict_in[str_key]["population"] = population_in
		dict_in[str_key]["date_mod"] = get_current_date_proc ()
	end

	return	dict_in
end
-- ----------------------------------------------------
function dict_delete_proc (dict_in,id_in)
	str_key = tostring (id_in)

	if (dict_in[str_key]) then
		dict_in[str_key] = nil
	end

	return	dict_in
end
-- ----------------------------------------------------
function dict_append_proc (dict_aa,id_in,name_in,population_in,date_mod_in)
	unit_aa = {name=name_in,population=population_in,date_mod=date_mod_in}
	dict_aa[id_in] = unit_aa 
	return dict_aa
end
-- ----------------------------------------------------
function csv_read_proc (file_in)
	fp_in, msg=io.open (file_in,"r")

	doc_aa = {}

	if (fp_in) then
		tt = fp_in:read ("*a")
		fp_in:close ()
		lines=split_proc (tt, "\n")
		for it=1 , #lines do
			rr=split_proc (lines[it], ",")
			if (2 < #rr) then
				doc_aa[rr[1]] =
				 {name= rr[2],population=rr[3],date_mod=rr[4]}
			end
		end
	else
		print (msg)
	end

	return	doc_aa
end

-- ----------------------------------------------------
function csv_write_proc (file_out,dict_in)
	io.output (file_out)
	for str_key in pairs (dict_in)
		do
			if (dict_in[str_key]) then
				out_str = join_proc (dict_in,str_key,",") .. "\n"
				io.write (out_str)
			end
		end
end
-- ----------------------------------------------------
function get_current_date_proc ()
	today = (os.date ("*t").year) .. "-" ..
		(os.date ("*t").month) .. "-" ..
		(os.date ("*t").day)
--
	return	today
end
-- ----------------------------------------------------
